### ISSUE TITLE
Don't crash when DB columns are reordered (e.g. when columns are added in the middle)

### ISSUE THREAD

**Issue by cmeeren** [fsprojects/FSharp.Data.SqlClient#383]
I just pushed a change to a DB where I added a column somewhere in the middle. This caused a number of InvalidCastExceptions from the existing running app. Once I published the updated code where SqlClient knows about the added field, everything worked fine again.

This seems to indicate that internally, SqlClient uses column indexes. Could you instead switch to using column names? That way, changing the column order of existing databases (such as when inserting or deleting columns) won't crash and burn existing apps.

**Comment by smoothdeveloper** [comment]
@cmeeren are you using 'select *' pattern in the query?

I avoid this pattern in production SQL query, I assume specifying the fields explicitly in the query would avoid the issue in the first place.

I'm ambivalent on fixing this versus keeping perspective `select *` as an anti pattern, I assume if column ordinal was used in the first place, there was a performance aspect to it.

You can see that it implies an extra call to GetOrdinal for every single value read out of the `SqlDataReader`: https://referencesource.microsoft.com/#system.data/system/data/sqlclient/SqlDataReader.cs,396

**Comment by cmeeren** [comment]
Thanks for getting back to me. Let's break it down.

## `SELECT *` being an anti-pattern

I almost always use `SELECT *`. AFAIK the reason it's viewed as a code smell is because you may fetch more data than you need. That is never the case for me, because all my tables are slim single-entity tables where I always need to fetch all columns in order to construct the domain entity. On the other hand, the benefit of `SELECT *` (and why people use it at all) is maintenance and clarity:
* I never have to update my "fetch entity" queries because they always fetch any new column I add to my domain entity tables (again, I generally always need them all; in the very few cases that I don't, I can take care of that).
* It's much easier to read the intention of `SELECT * FROM Company where CompanyId = @companyId` ("fetch this entire company") rather than if I used a long list of explicit columns (you'd have to study the list thoroughly to see that yes, you did indeed include all columns, making this a "fetch this entire company" query instead of a "fetch parts but for some reason not all of the company" query).

## Working around the problem by specifying order in `SELECT` queries

Based on your explanation, it seems likely that this would work around the problem. However, I have three objections here:
* This only moves the problem from the table definition to the SELECT query.
* I like to order the output column list in my SELECT queries for clarity.
* **Most importantly: You don't get compile-safety for the SELECT order.** If you get the order wrong when expanding SELECT queries, nothing will tell you that until you either get the `InvalidCastException` as I described, or even worse, you silently get the wrong data in the wrong fields (e.g. load the user's first name into the domain entity's last name field because you changed the SELECT order).

## Performance impacts

As with anything involving IO, my expectation is that if you measure it, this extra call to `GetOrdinal` would be an order of magnitude (or more) faster than actually fetching the data. The comments to [this SO answer](https://stackoverflow.com/a/1079382/2978652) seems to support this. If you want to use column names, you have four options:

* Ignore the above and use the indexer each time. I think this is fine.
* Same as the above, but also keep the current behavior and use a type provider parameter to let users switch between ordinal and column name.
* Call `GetOrdinal` manually once for all columns, and cache the values at runtime, meaning you only call `GetOrdinal` for the first row, as explained the SO anwer linked above. I think this (or the below) option is the ideal one.
* Statically generate the correct ordinal mapping at compile-time if possible.

In summary: Given that F# and type providers are all about compile-time safety, I think the current behavior is dangerous. If you want to keep this behavior, IMHO it should be very clearly documented that the order of SELECT columns in sprocs must be stable and new columns only appended at the end. But as I have said, that can give messy/confusing SELECT column lists, and isn't compile-time safe anyway.

The change seems fairly simple code-wise (and non-breaking). Please let me know if you want help with anything, or are confused about or disagree with some of my points.

**Comment by smoothdeveloper** [comment]
@cmeeren I forgot to ask if you are using the `DataTable` result type or if it had any impact in your case if you are currently using `Record` or `Tuple`.

For the points on `select *`, I'm not following/understanding the objections in context of the TP as it is totally compile time safe when the columns are stated explicitly, while it may not be the case otherwise, whenever there is inconsistency of column order between development and production databases.

My usage doesn't follow same idioms, there are most likely joins in my queries, and fields with same name may appear in different tables, aliases may be used in any of the columns, etc.

`select *` also has issues when used with `insert` statement (where you can also omit the column list, which is similar anti-pattern), also, when a schema evolves, you can't easily guarantee the order of fields is the same between a database created from scratch (new columns may be inserted not at the end) and one which has been patched (new columns are always last).

Those are reasons I always encourage the explicit list of columns as saner default, albeit a bit verbose and requiring to adjust few queries once in a while.

For the performance of named versus ordinal indexing, this point came to mind because performance is being discussed on http://fsprojects.github.io/FSharp.Data.SqlClient/comparison.html and AFAIU, the choices made in the ADO code generated by the TP seems to always gear toward the low-level approach.

If the impact was tested on significant dataset (millions of values) and there is no measurable performance impact, changing the implementation seems fine.

For the guarantee you are after, I'd expect an implementation to come with the tests showcasing it to avoid any regression.

It would also be interesting to see what is the status in other mapping libraries to know if they fail the same way or deal with it better.

**Comment by cmeeren** [comment]
> @cmeeren I forgot to ask if you are using the `DataTable` result type or if it had any impact in your case if you are currently using `Record` or `Tuple`.

The default, which is `Record` I guess.

> For the points on `select *`, I'm not following/understanding the objections in context of the TP as it is totally compile time safe when the columns are stated explicitly, **while it may not be the case otherwise, whenever there is inconsistency of column order between development and production databases.**

(Emphasis mine) Yes, that's it. This whole issue is about what happens during DB migrations. You don't even need separate dev/prod databases. When you publish a new version of the app, then first the DB must be migrated, and then the new app published. This means that **all DB migrations must be backwards-compatible with the currently running version of the app**, otherwise the currently running instance will break between publishing the DB and publishing the app. (Note that I publish the DB as a separate first step using DACPAC publish; if you publish using the app startup code, things may be different.)

My problem here is that I expected adding a new (nullable) column to a table to be backwards compatible no matter where it's added, but it turned out it wasn't, because then the currently running version of the app received the columns in another order. This failed because the column index was used. It would not fail if the column name was used, which would make these kinds of changes backwards compatible.

Again, you can work around it by using explicit `SELECT Col1, Col2, ...` in the sproc, but then you have the same problem there: You can't reorder the `SELECT` fields because that will break for the same reason.

> My usage doesn't follow same idioms, there are most likely joins in my queries, and fields with same name may appear in different tables, aliases may be used in any of the columns, etc.

I too use joins, but then I qualify it using the table name: `SELECT Company.* FROM Company INNER JOIN SomeFilteringTable ...`.

> `select *` also has issues when used with `insert` statement (where you can also omit the column list, which is similar anti-pattern)

I agree that `insert` should specify explicit fields for the reason you mention. But I don't understand what `SELECT *` has to do with this.

> also, when a schema evolves, you can't easily guarantee the order of fields is the same between a database created from scratch (new columns may be inserted not at the end) and one which has been patched (new columns are always last).

That's my point; I don't want to care about the order. If I have an sproc that contains

```sql
SELECT FirstName, Phone, LastName FROM [User] ...
```

I want to be able to reorder the fields for improved readability:

```sql
SELECT FirstName, LastName, Phone FROM [User] ...
```

(The above example is of course a bit artificial.)

If I make the change above, with the way SqlClient currently works, this DB migration (updating the sproc) will break the currently running app.

> Those are reasons I always encourage the explicit list of columns as saner default, albeit a bit verbose and requiring to adjust few queries once in a while.

I get that it's a good idea. I'm not strongly opposed to explicitly specifying the list (though I prefer `*`). My main problem here is that, as explained above, even if I do that, I cannot reorder the fields in the `SELECT` list in an sproc without breaking the currently running app.

> For the guarantee you are after, I'd expect an implementation to come with the tests showcasing it to avoid any regression.

I don't know anything about testing TPs, but I guess that the following would reproduce it, I hope it's helpful:
* Define an sproc that selects some columns from a table, publish to DB
* Use `SqlProgrammabilityProvider` to call the sproc, everything should work
* Now change the sproc so that the `SELECT` fields are re-ordered (preferably with incompatible types in the old ordinal positions)
* Run the query *without recompiling*, it should now fail with `InvalidCastException`

Again, I have no idea how to test this since I don't know anything about testing TPs, but perhaps you know.

**Comment by cmeeren** [comment]
To potentially clarify, AFAIK the problem I experience is only relevant if you use sprocs. If you use inline or `SqlFile` queries with `SqlCommandProvider`, then if course there is no migration in the DB that changes the field order of queries, and the problem can be avoided by explicitly specifying the `SELECT` fields (you are then free to reorder them without breaking anything). Though `SELECT *` will still cause problems.

My setup is centered around a DB project in VS that contains my schema, and I define sprocs there and call the sprocs using `SqlProgrammabilityProvider` because sproc development in a DB project is superior to SQL script development outside of a DB project. In particular:
* The intellisense experience is much better. Intellisense for script development requires you to connect to a DB every single time you open a script, which typically takes 10 seconds because the VS UI for this is slow.
* There's no need for all the verbose, hacky workarounds mentioned in [Dynamic creation of offline MDF](http://fsprojects.github.io/FSharp.Data.SqlClient/dynamic%20local%20db.html). Simply deploy the DB project to a localhost DB and point SqlClient to that DB at compile time.
* Very simple and robust migrations using DACPACs.

**Comment by smoothdeveloper** [comment]
@cmeeren, I'm closing this, I remain of the opinion `select *` should be avoided, like few other anti patterns.

If you feel it is not reasonable, please provide an implementation that would satisfy your requirements along with test case ensuring it doesn't regress.

**Comment by cmeeren** [comment]
> I remain of the opinion `select *` should be avoided, like few other anti patterns.

I cannot make my case clearer than I already have, so I guess we just disagree then. :) Could I ask you what parts of my reasoning you are unconvinced by?

> If you feel it is not reasonable, please provide an implementation that would satisfy your requirements along with test case ensuring it doesn't regress.

Unfortunately, as mentioned, I have no idea how to test this.

In any case, I am considering sidestepping this and other issues completely by ditching TPs and instead writing a (hopefully) simple code generator that generates code for sprocs and SQL scripts, as a dotnet tool to be run as part of the build process. Easier to develop in itself, better IDE performance, inherently cacheable, easy to control when it regenerates (e.g. delete or edit the generated file), easier to debug, can in a pinch (critical production error) fix the generated code directly before submitting a bug report and waiting for a fix, etc. TPs are good for interactive exploration and large/infinite data spaces, but for executing a handful of sprocs or scripts, I think a code generator can be better.

**Comment by smoothdeveloper** [comment]
> Could I ask you what parts of my reasoning you are unconvinced by?

> *is because you may fetch more data than you need.  That is never the case for me*

When you faced the issue, your application binary was actually fetching more data than needed for it to run, also I think it is good to have deterministic order.

Since this is minor breakage due to updating database without updating the binaries, and it requires fair amount of work to do the change, I just said that it will be up for grabs if you really need that.

> Unfortunately, as mentioned, I have no idea how to test this.

The steps to reproduce your issue is good test for it, it needs to be put under the test suite to make sure that the behaviour would remain.

> I think a code generator can be better in almost all respects.

IMHO, it would be possible to contribute refactors to this library which is already dealing with lots of corner cases, has fair amount of logic to determine the types of generated code, all of that logic could be used from external tools.

Of course, the appeal of a standalone and done from scratch tool may be greater, please let us know if you are engaging in such efforts.

Thanks.

**Comment by cmeeren** [comment]
> IMHO, it would be possible to contribute refactors to this library

I'm trying :) #348

**Comment by smoothdeveloper** [comment]
Ok, I'll get through it, thanks for the reminder.
